### PR TITLE
Read etcd location from ETCD_ENDPOINTS if available

### PR DESCRIPTION
**Pull request by matthewdupre**

### CODE REVIEW

File: calico_containers/pycalico/datastore.py [hunk 1]
```diff
@@ -29,6 +29,8 @@
 
 ETCD_AUTHORITY_DEFAULT = "127.0.0.1:2379"
 ETCD_AUTHORITY_ENV = "ETCD_AUTHORITY"
+ETCD_ENDPOINTS_DEFAULT = ""
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py]: I don't think you need this

File: calico_containers/pycalico/datastore.py [hunk 2]
```diff
@@ -137,17 +139,55 @@ class DatastoreClient(object):
     """
 
     def __init__(self):
+        etcd_endpoints = os.getenv(ETCD_ENDPOINTS_ENV, ETCD_ENDPOINTS_DEFAULT)
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py]: Can't you just use the default default of None?

File: calico_containers/pycalico/datastore.py [hunk 3]
```diff
@@ -137,17 +139,55 @@ class DatastoreClient(object):
...
         etcd_cert = os.getenv(ETCD_CERT_FILE_ENV, '')
         etcd_ca = os.getenv(ETCD_CA_CERT_FILE_ENV, '')
+
+        etcd_addrs_str = []
+        if etcd_endpoints != ETCD_ENDPOINTS_DEFAULT:
+            # ETCD_ENDPOINTS specified: use it to determine scheme and etcd
+            # location.
+            endpoints = [x.strip() for x in etcd_endpoints.split(",")]
+            try:
+                scheme = None
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py]: Is it a str or a list?

File: calico_containers/pycalico/datastore.py [hunk 4]
```diff
@@ -137,17 +139,55 @@ class DatastoreClient(object):
...
-                                 "<address>:<port>. Value provided is '%s'" %
-                                 (ETCD_AUTHORITY_ENV, etcd_authority))
-
-        (host, port) = etcd_authority.split(":", 1)
         etcd_scheme = os.getenv(ETCD_SCHEME_ENV, ETCD_SCHEME_DEFAULT)
         etcd_key = os.getenv(ETCD_KEY_FILE_ENV, '')
         etcd_cert = os.getenv(ETCD_CERT_FILE_ENV, '')
         etcd_ca = os.getenv(ETCD_CA_CERT_FILE_ENV, '')
+
+        etcd_addrs_str = []
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py]: Is it a str or a list?

File: calico_containers/pycalico/datastore.py [hunk 5]
```diff
@@ -137,17 +139,55 @@ class DatastoreClient(object):
...
+        if etcd_endpoints != ETCD_ENDPOINTS_DEFAULT:
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py]: You can drop the != ETCD_ENDPOINTS_DEFAULT bit

File: calico_containers/pycalico/datastore.py [hunk 6]
```diff
@@ -137,17 +139,55 @@ class DatastoreClient(object):
...
+            # ETCD_ENDPOINTS specified: use it to determine scheme and etcd
+            # location.
+            endpoints = [x.strip() for x in etcd_endpoints.split(",")]
+            try:
+                scheme = None
+                for e in endpoints:
+                    s, a = e.split("://")
+                    etcd_addrs_str.append(a)
+                    if scheme == None:
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py:160]: Why not just set etcd_scheme = s and drop this if test and the assignment below?

Comment (tomdee) [inline calico_containers/pycalico/datastore.py:160]: Oh, I see, so you can detect the inconsistent protocols

File: calico_containers/pycalico/datastore.py [hunk 7]
```diff
@@ -186,11 +226,21 @@ def __init__(self):
...
-                                       ca_cert=etcd_ca)
+
+        # python-etcd Client requires a different invocation when there's only
+        # a single etcd host.
+        if len(etcd_addrs) > 1:
+            self.etcd_client = etcd.Client(host=tuple(etcd_addrs),
+                                           protocol=etcd_scheme,
+                                           cert=key_pair,
+                                           ca_cert=etcd_ca,
+                                           allow_reconnect=True)
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py:244]: Is it worth a note on the allow_reconnects?

File: calico_containers/pycalico/datastore.py [hunk 8]
```diff
@@ -137,17 +139,55 @@ class DatastoreClient(object):
...
+            # ETCD_SCHEME instead.
+            etcd_addrs_str.append(etcd_authority)
+
+        etcd_addrs = []
+        for addr in etcd_addrs_str:
+            if not validate_hostname_port(addr):
+                raise DataStoreError(
+                    "Invalid %s. It must take the form "
+                    "<address>:<port>. Value provided is '%s'" %
+                    (ETCD_AUTHORITY_ENV, etcd_authority)
```
Comment (tomdee) [inline calico_containers/pycalico/datastore.py]: This error will be wrong
1) if the data came from ETCD_ENDPOINTS
2) And the value it's logging should be addr and not etcd_authority

**Comment by tomdee** [comment]
I've finished looking at this - can you apply the markups and I'll take another look.

**Comment by tomdee** [comment]
LGTM - please squash and merge

File: calico_containers/pycalico/datastore.py (hunk 2 above)
Comment (matthewdupre) [inline calico_containers/pycalico/datastore.py]: I've gone with what matches the other calls a few lines below (and throughout most of the rest of Calico), to which the same argument could apply.

File: calico_containers/pycalico/datastore.py (hunk 4 above)
Comment (matthewdupre) [inline calico_containers/pycalico/datastore.py]: Renamed to raw: it's actually a list of strings.

File: calico_containers/pycalico/datastore.py (hunk 1 above)
Comment (matthewdupre) [inline calico_containers/pycalico/datastore.py]: Removed.

File: calico_containers/pycalico/datastore.py (hunk 5 above)
Comment (matthewdupre) [inline calico_containers/pycalico/datastore.py]: Removed.

File: calico_containers/pycalico/datastore.py (hunk 7 above)
Comment (matthewdupre) [inline calico_containers/pycalico/datastore.py:244]: Added.

**Pull request merged**